<a href="https://colab.research.google.com/github/SuperNZH/KPMG_DataAnalytics_Virtual_Program/blob/main/1_Data_Quality_Assessment/data_quality_assessment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Mount Drive and Data Reading

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

In [ ]:
path = "/content/drive/MyDrive/Colab_Notebooks/my_project/kpmg_virtual_intern/kpmg.xlsx"
transactions = pd.read_excel(path, sheet_name=1, header=1)
new_customer_list = pd.read_excel(path, sheet_name=2, header=1)
customer_demographic = pd.read_excel(path, sheet_name=3, header=1)
customer_address = pd.read_excel(path, sheet_name=4, header=1)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype={value.dtype})
  This is separate from the ipykernel package so we can avoid doing imports until
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: FutureWarning: Inferring datetime64[ns] from data containing strings is deprecated and will be removed in a future version. To retain the old behavior explicitly pass Series(data, dtype={value.dtype})
  after removing the cwd from sys.path.


### Data Sample

In [ ]:
transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0             1.0         2.0       2950.0       2017-02-25           0.0   
1             2.0         3.0       3120.0       2017-05-21           1.0   
2             3.0        37.0        402.0       2017-10-16           0.0   
3             4.0        88.0       3135.0       2017-08-31           0.0   
4             5.0        78.0        787.0       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [ ]:
new_customer_list.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                 6.0        0.83      1.0375     1.296875     1.102344   
1                11.0        0.55      0.5500     0.687500     0.584375   
2                 5.0        1.10      1.1000     1.100000     1.100000   
3                 1.0        0.69      0.8625     0.862500     0.862500   
4                 9.0        0.99      0.9900     1.237500     1.237500   

   Unnamed: 20  Rank     Value  
0            1   1.0  1.718750  
1            1   1.0  1.718750  
2            1   1.0  1.718750  
3            4   4.0  1.703125  
4            4   4.0  1.703125  

[5 rows x 23 columns]

In [ ]:
customer_demographic.head()

customer_id      first_name  last_name  gender  \
0          1.0         Laraine  Medendorp       F   
1          2.0             Eli    Bockman    Male   
2          3.0           Arlin     Dearle    Male   
3          4.0          Talbot        NaN    Male   
4          5.0  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                 93.0 1953-10-12     Executive Secretary   
1                                 81.0 1980-12-16  Administrative Officer   
2                                 61.0 1954-01-20      Recruiting Manager   
3                                 33.0 1961-10-03                     NaN   
4                                 56.0 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [ ]:
customer_address.head()

customer_id              address  postcode            state    country  \
0          1.0   060 Morning Avenue    2016.0  New South Wales  Australia   
1          2.0  6 Meadow Vale Court    2153.0  New South Wales  Australia   
2          4.0   0 Holy Cross Court    4211.0              QLD  Australia   
3          5.0  17979 Del Mar Point    2448.0  New South Wales  Australia   
4          6.0     9 Oakridge Court    3216.0              VIC  Australia   

   property_valuation  
0                10.0  
1                10.0  
2                 9.0  
3                 4.0  
4                 9.0

## Transactions Table

In [ ]:
trans_null = transactions.isnull().sum()
trans_null

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

**Featues with Missing Values**:

online_order(360) and brand, product_line, product_class, 
product_size(197)

In [ ]:
transactions.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

In [ ]:
# Use this line and change the column name to check the abnormal values 
transactions.sort_values(by=['transaction_date'], ascending=False)

transaction_id  product_id  customer_id transaction_date  online_order  \
1696           1697.0        56.0       2421.0       2017-12-30           1.0   
15269         15270.0         3.0        195.0       2017-12-30           0.0   
4668           4669.0        29.0       1052.0       2017-12-30           0.0   
13574         13575.0        98.0       2396.0       2017-12-30           0.0   
3937           3938.0        95.0        993.0       2017-12-30           0.0   
...               ...         ...          ...              ...           ...   
16397         16398.0        43.0       2160.0       2017-01-01           1.0   
15204         15205.0        87.0        123.0       2017-01-01           0.0   
15137         15138.0         3.0       1022.0       2017-01-01           1.0   
3459           3460.0        64.0        234.0       2017-01-01           1.0   
13933         13934.0        94.0        824.0       2017-01-01           0.0   

      order_status           brand product_line product_class product_size  \
1696      Approved      OHM Cycles     Standard        medium       medium   
15269     Approved   Trek Bicycles     Standard        medium        large   
4668      Approved        WeareA2B     Standard        medium       medium   
13574     Approved   Trek Bicycles     Standard          high       medium   
3937      Approved      OHM Cycles      Touring           low       medium   
...            ...             ...          ...           ...          ...   
16397     Approved           Solex     Standard        medium       medium   
15204     Approved      OHM Cycles     Standard        medium       medium   
15137     Approved   Trek Bicycles     Standard        medium        large   
3459      Approved  Giant Bicycles     Standard          high        small   
13933     Approved  Giant Bicycles     Standard        medium        large   

       list_price  standard_cost  product_first_sold_date  
1696       183.86         137.90                  33259.0  
15269     2091.47         388.92                  41167.0  
4668      1065.03         230.09                  36833.0  
13574      358.39         215.03                  33364.0  
3937      1073.07         933.84                  35455.0  
...           ...            ...                      ...  
16397     1151.96         649.49                  41009.0  
15204     1636.90          44.71                  40410.0  
15137     2091.47         388.92                  36367.0  
3459      1977.36        1759.85                  40779.0  
13933     1635.30         993.66                  38002.0  

[20000 rows x 13 columns]

The main problems of transactions table is missing values

There are 5 columns in transactions table with missing values, and below is the info:

* online_order ->               360
* brand   ->                    197
* product_line ->               197
* product_class    ->           197
* product_size     ->           197
* standard_cost       ->        197
* product_first_sold_date  ->   197

## NewCustomerList Table

In [ ]:
ncl_null = transactions.isnull().sum()
ncl_null

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

Columns with Missing values in the newCustomerList


* online_order               360
* brand                      197
* product_line               197
* product_class              197
* product_size               197
* standard_cost              197
* product_first_sold_date    197

In [ ]:
new_customer_list.describe()

past_3_years_bike_related_purchases       tenure     postcode  \
count                          1000.000000  1000.000000  1000.000000   
mean                             49.836000    11.388000  3019.227000   
std                              27.796686     5.037145   848.895767   
min                               0.000000     0.000000  2000.000000   
25%                              26.750000     7.000000  2209.000000   
50%                              51.000000    11.000000  2800.000000   
75%                              72.000000    15.000000  3845.500000   
max                              99.000000    22.000000  4879.000000   

       property_valuation  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  \
count         1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean             7.397000     0.740390     0.832140     0.935666     0.863405   
std              2.758804     0.203073     0.248295     0.298383     0.281441   
min              1.000000     0.400000     0.400000     0.400000     0.340000   
25%              6.000000     0.570000     0.620000     0.700000     0.648125   
50%              8.000000     0.730000     0.825000     0.912500     0.840000   
75%              9.000000     0.920000     1.010000     1.137500     1.037500   
max             12.000000     1.100000     1.375000     1.718750     1.718750   

       Unnamed: 20         Rank        Value  
count  1000.000000  1000.000000  1000.000000  
mean    498.819000   498.819000     0.881714  
std     288.810997   288.810997     0.293525  
min       1.000000     1.000000     0.340000  
25%     250.000000   250.000000     0.649531  
50%     500.000000   500.000000     0.860000  
75%     750.250000   750.250000     1.075000  
max    1000.000000  1000.000000     1.718750

In [ ]:
# Use this line and change the column name to check the abnormal values 
new_customer_list.sort_values(by=['DOB'], ascending=True)

first_name   last_name  gender  past_3_years_bike_related_purchases  \
797    Anthony     Morison    Male                                    7   
885  Zachariah  Meininking    Male                                   40   
769     Andrea      Pendle  Female                                   86   
585    Letizia       Poore  Female                                   27   
265     Alleen       Eaken  Female                                   56   
..         ...         ...     ...                                  ...   
775     Dmitri       Viant       U                                   62   
835      Porty      Hansed       U                                   88   
883      Shara    Bramhill       U                                   24   
904       Roth        Crum       U                                    0   
984    Pauline    Dallosso       U                                   82   

           DOB                   job_title job_industry_category  \
797 1938-06-08             General Manager                Health   
885 1938-06-09             Cost Accountant    Financial Services   
769 1938-08-05                         NaN                   NaN   
585 1938-08-30            Web Developer II         Manufacturing   
265 1938-08-31                         NaN                   NaN   
..         ...                         ...                   ...   
775        NaT                   Paralegal    Financial Services   
835        NaT             General Manager                    IT   
883        NaT                         NaN                    IT   
904        NaT             Legal Assistant                    IT   
984        NaT  Desktop Support Technician                    IT   

        wealth_segment deceased_indicator owns_car  ...  state    country  \
797      Mass Customer                  N       No  ...    NSW  Australia   
885  Affluent Customer                  N      Yes  ...    NSW  Australia   
769     High Net Worth                  N      Yes  ...    QLD  Australia   
585  Affluent Customer                  N       No  ...    VIC  Australia   
265      Mass Customer                  N       No  ...    NSW  Australia   
..                 ...                ...      ...  ...    ...        ...   
775  Affluent Customer                  N       No  ...    VIC  Australia   
835      Mass Customer                  N       No  ...    NSW  Australia   
883  Affluent Customer                  N       No  ...    NSW  Australia   
904      Mass Customer                  N       No  ...    NSW  Australia   
984  Affluent Customer                  N      Yes  ...    NSW  Australia   

     property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
797                 9.0        0.60      0.6000      0.75000     0.637500   
885                 7.0        1.01      1.2625      1.26250     1.262500   
769                 6.0        0.66      0.8250      1.03125     1.031250   
585                 3.0        0.52      0.5200      0.52000     0.520000   
265                12.0        0.91      0.9100      1.13750     0.966875   
..                  ...         ...         ...          ...          ...   
775                 1.0        0.70      0.7000      0.70000     0.700000   
835                11.0        0.51      0.5100      0.63750     0.541875   
883                10.0        0.58      0.5800      0.72500     0.725000   
904                 6.0        0.85      0.8500      0.85000     0.722500   
984                10.0        0.90      1.1250      1.40625     1.406250   

     Unnamed: 20   Rank     Value  
797          797  797.0  0.600000  
885          883  883.0  0.531250  
769          760  760.0  0.637500  
585          583  583.0  0.787500  
265          259  259.0  1.062500  
..           ...    ...       ...  
775          774  774.0  0.626875  
835          832  832.0  0.575000  
883          883  883.0  0.531250  
904          904  904.0  0.500000  
984          985  985.0  0.408000  

[10

## CustomerDemographic Table

In [ ]:
cd_null = customer_demographic.isnull().sum()
cd_null

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

Features with missing values

* last_name                              125
* DOB                                     87
* job_title                              506
* job_industry_category                  656
* default                                302
* tenure                                  87

In [ ]:
customer_demographic['wealth_segment'].value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

In [ ]:
customer_demographic['default'].value_counts()

100.0                                     113
1.0                                       112
-1.0                                      111
-100.0                                     99
١٢٣                                     53
                                         ... 
testâ testâ«                               31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: default, Length: 90, dtype: int64

In [ ]:
customer_demographic['deceased_indicator'].value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

## CustomerAddress Table

In [ ]:
ca_null = customer_demographic.isnull().sum()
ca_null

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64